### PR TITLE
[SPARK-37527][SQL] Translate more standard aggregate functions for pushdown

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, Spark aggregate pushdown will translate some standard aggregate functions, so that compile these functions to adapt specify database.
After this job, users could override `JdbcDialect.compileAggregate` to implement some aggregate functions supported by some database.
Because some aggregate functions will be converted show below, this PR no need to match them.

|Input|Parsed|Optimized|
|------|--------------------|----------|
|`Every`| `aggregate.BoolAnd` |`Min`|
|`Any`| `aggregate.BoolOr` |`Max`|
|`Some`| `aggregate.BoolOr` |`Max`|

### Why are the changes needed?
Make the implement of `*Dialect` could extends the aggregate functions by override `JdbcDialect.compileAggregate`.

### Does this PR introduce _any_ user-facing change?
Yes. Users could pushdown more aggregate functions.

### How was this patch tested?
Exists tests.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/AnyOrSome.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.expressions.NamedReference;
+
+/**
+ * An aggregate function that returns true if at least one value of `column` is true.
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public final class AnyOrSome implements AggregateFunc {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/AnyOrSome.java:29]: ANY/SOME will be replaced by MAX/MIN in spark, so we can't really hit in at runtime, and no need to add data source push down API for it.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Corr.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+public final class Corr implements AggregateFunc {
+    private final NamedReference x;
+    private final NamedReference y;
+
+    public Corr(NamedReference left, NamedReference right) {
+        this.x = left;
+        this.y = right;
+    }
+
+    public NamedReference getX() { return x; }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Corr.java:39]: let's follow other aggregate functions and use scala style. Just name it `x`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/CovarPop.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * An aggregate function that returns the population covariance of a set of number pairs.
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public final class CovarPop implements AggregateFunc {
+    private final NamedReference left;
+    private final NamedReference right;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/CovarPop.java:31]: why not `x` and `y`?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Every.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.expressions.NamedReference;
+
+/**
+ * An aggregate function that returns true if all values of `column` are true.
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public final class Every implements AggregateFunc {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Every.java:29]: ditto, please check if we can really see it in the physical plan

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Corr.java (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/Corr.java:39]: Thank you for the reminder.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/CovarPop.java (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/aggregate/CovarPop.java:31]: OK

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 5]
```diff
@@ -208,6 +208,9 @@ abstract class JdbcDialect extends Serializable with Logging{
       case max: Max =>
         if (max.column.fieldNames.length != 1) return None
         Some(s"MAX(${quoteIdentifier(max.column.fieldNames.head)})")
+      case avg: Average =>
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:211]: You means pushdown `sum/count` to data source ? Why not use `avg` ?

**Comment by beliefer** [comment]
ping @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:211]: In Spark, partial aggregate output for avg is a sequence of (sum, count). If we want to push down partial aggregate of avg to data source, should we also use sum and count?

[the remark this answers is not in the document]
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:211]: But for final aggregate, it just need the value of avg(na matter it is partial aggregate or pushdown aggregate).

**Comment by beliefer** [comment]
https://github.com/apache/spark/pull/35101 merged.

**Pull request closed without merging**